tap_azure_tickets/__init__.py

```diff
@@ -274,19 +250,51 @@ def request(source, url, method='GET', json=None):
-        timer.tags['url'] = url
-
-        response = session.request(method=method, url=url, json=json)
-
-        timer.tags[metrics.Tag.http_status_code] = response.status_code
-        timer.tags['header_retry-after'] = response.headers.get('retry-after')
-        timer.tags['header_x-ratelimit-resource'] = response.headers.get('x-ratelimit-resource')
-        timer.tags['header_x-ratelimit-delay'] = response.headers.get('x-ratelimit-delay')
-        timer.tags['header_x-ratelimit-limit'] = response.headers.get('x-ratelimit-limit')
-        timer.tags['header_x-ratelimit-remaining'] = response.headers.get('x-ratelimit-remaining')
-        timer.tags['header_x-ratelimit-reset'] = response.headers.get('x-ratelimit-reset')
-        return response
+
+    exponential_factor = 5
+    tries = 0
+    def backoff_value(response):
+        nonlocal tries
+        with suppress(TypeError, ValueError, AttributeError):
+            return int(response.headers.get("Retry-After"))
+        backoff_time = exponential_factor * (2 ** tries)
+        tries += 1
+        return backoff_time
+
+    def execute_request(source, url, method='GET', json=None):
+        with metrics.http_request_timer(source) as timer:
+            timer.tags['url'] = url
+
+            response = session.request(method=method, url=url, json=json)
+
+            timer.tags[metrics.Tag.http_status_code] = response.status_code
+            timer.tags['header_retry-after'] = response.headers.get('retry-after')
+            timer.tags['header_x-ratelimit-resource'] = response.headers.get('x-ratelimit-resource')
+            timer.tags['header_x-ratelimit-delay'] = response.headers.get('x-ratelimit-delay')
+            timer.tags['header_x-ratelimit-limit'] = response.headers.get('x-ratelimit-limit')
+            timer.tags['header_x-ratelimit-remaining'] = response.headers.get('x-ratelimit-remaining')
+            timer.tags['header_x-ratelimit-reset'] = response.headers.get('x-ratelimit-reset')
+
+            return response
+
+    backoff_on_exception = backoff.on_exception(backoff.expo,
+                      (requests.exceptions.RequestException),
+                      max_tries=7,
+                      max_time=3600,
+                      giveup=lambda e: e.response is not None and e.response.status_code != 429 and 400 <= e.response.status_code < 500,
+                      factor=exponential_factor,
+                      jitter=backoff.random_jitter,
+                      logger=logger)
+
+    backoff_on_predicate = backoff.on_predicate(backoff.runtime,
+                      predicate=lambda r: r.headers.get("Retry-After", None) != None or r.status_code >= 300,
+                      max_tries=7,
+                      max_time=3600,
+                      value=backoff_value,
+                      jitter=backoff.random_jitter,
+                      logger=logger)
+
+    return backoff_on_exception(backoff_on_predicate(execute_request))(source, url, method, json)
```

Comment:
Just like decorators wrap functions, these functions also work the same way.  So, execute_request ultimately gets all the variables passed into it, but the syntax looks a little funny.